portfolio/src/test/java/com/google/sps/DataServletTest.java

```diff
@@ -56,127 +64,232 @@ public void tearDown() {
-    new DataServlet().doPost(postRequest, postResponse);
+    dataServelet.doPost(postRequest, postResponse);
     when(postRequest.getParameter("text-input")).thenReturn("Test Comment #3");
-    new DataServlet().doPost(postRequest, postResponse);
-    when(getRequest.getParameter("max-num")).thenReturn("3");
-    new DataServlet().doGet(getRequest, getResponse);
+    dataServelet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter("text-input")).thenReturn("Test Comment #4");
+    dataServelet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter("text-input")).thenReturn("Test Comment #5");
+    dataServelet.doPost(postRequest, postResponse);
+    when(getRequest.getParameter("max-num")).thenReturn(Integer.toString(NUM_COMMENTS));
+    dataServelet.doGet(getRequest, getResponse);
 
     printWriter.flush(); //may not have flushed yet
-    assertThat(stringWriter.toString()).contains("Test Comment #1");
-    assertThat(stringWriter.toString()).contains("Test Comment #2");
-    assertThat(stringWriter.toString()).contains("Test Comment #3");
+    String[] comments = convertStringToArray(stringWriter.toString());
+    assertThat(comments[0]).isEqualTo("Test Comment #1");
+    assertThat(comments[1]).isEqualTo("Test Comment #2");
+    assertThat(comments.length).isEqualTo(2);
   }
 
   @Test
-  public void testGETEmptyDataStore() throws Exception {
-    HttpServletRequest getRequest = mock(HttpServletRequest.class);
-    HttpServletResponse getResponse = mock(HttpServletResponse.class);
+  public void testPostAndGetAllComments() throws Exception {
+    int NUM_COMMENTS = 3;
 
     StringWriter stringWriter = new StringWriter();
     PrintWriter printWriter = new PrintWriter(stringWriter);
-    when(getResponse.getWriter()).thenReturn(printWriter);  
+    when(getResponse.getWriter()).thenReturn(printWriter);
 
-    new DataServlet().doGet(getRequest, getResponse); 
+    when(postRequest.getParameter("text-input")).thenReturn("Test Comment #1");
+    dataServelet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter("text-input")).thenReturn("Test Comment #2");
+    dataServelet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter("text-input")).thenReturn("Test Comment #3");
+    dataServelet.doPost(postRequest, postResponse);
+    when(getRequest.getParameter("max-num")).thenReturn(Integer.toString(NUM_COMMENTS));
+    dataServelet.doGet(getRequest, getResponse);
 
-    printWriter.flush(); //may not have flushed yets
-    assertThat(stringWriter.toString().trim()).isEqualTo("[]");
+    printWriter.flush(); //may not have flushed yet
+    String[] comments = convertStringToArray(stringWriter.toString());
+    assertThat(comments[0]).isEqualTo("Test Comment #1");
+    assertThat(comments[1]).isEqualTo("Test Comment #2");
+    assertThat(comments[2]).isEqualTo("Test Comment #3");
+    assertThat(comments.length).isEqualTo(3);
```

Comment:
I think these three tests are largely redundant with the tests that you've already written that isolate specific bits of functionality in the class under test. 

I do like the idea of these tests (and I'm glad you wrote them). I think these would be better expressed as integration tests that actually fully exercise the system. For now, why don't we just keep `testPostAndGetSingleComment` as a unit test. We can follow up offline to discuss testing strategies.

Reply:
sounds good!